src/gui/torrentcontentmodel.cpp

```diff
@@ -81,44 +86,80 @@ namespace
             return cached;
         }
     };
-#if defined(Q_OS_WIN)
-    // See QTBUG-25319 for explanation why this is required
-    class WinShellFileIconProvider: public UnifiedFileIconProvider
+
+#ifdef QBT_PIXMAP_CACHE_FOR_FILE_ICONS
+    Q_DECL_UNUSED struct PixmapCacheSetup
+    {
+        static const int PixmapCacheForIconsSize = 2 * 1024 * 1024; // 2 MiB for file icons
+
+        PixmapCacheSetup()
+        {
+            QPixmapCache::setCacheLimit(QPixmapCache::cacheLimit() + PixmapCacheForIconsSize);
+        }
+
+        ~PixmapCacheSetup()
+        {
+            Q_ASSERT(QPixmapCache::cacheLimit() > PixmapCacheForIconsSize);
+            QPixmapCache::setCacheLimit(QPixmapCache::cacheLimit() - PixmapCacheForIconsSize);
+        }
+    };
+
+    PixmapCacheSetup pixmapCacheSetup;
+
+    class CachingFileIconProvider: public UnifiedFileIconProvider
     {
     public:
         using QFileIconProvider::icon;
 
-        QIcon icon(const QFileInfo &info) const override
+        QIcon icon(const QFileInfo &info) const final
         {
+            const QString ext = info.suffix();
+            if (!ext.isEmpty()) {
+                QPixmap cached;
+                if (QPixmapCache::find(ext, &cached)) return QIcon(cached);
+
+                const QPixmap pixmap = pixmapForExtension(ext);
+                if (!pixmap.isNull()) {
+                    QPixmapCache::insert(ext, pixmap);
+                    return QIcon(pixmap);
+                }
+            }
+            return UnifiedFileIconProvider::icon(info);
+        }
+
+    protected:
+        virtual QPixmap pixmapForExtension(const QString &ext) const = 0;
+    };
+#endif
+
+#if defined(Q_OS_WIN)
+    // See QTBUG-25319 for explanation why this is required
+    class WinShellFileIconProvider final: public CachingFileIconProvider
+    {
+    protected:
+        QPixmap pixmapForExtension(const QString &ext) const final override
```

Comment:
All members of final class are final, isn't it? So final here is redundant space waster.